Jint/Engine.cs

```diff
@@ -162,6 +166,54 @@ public Engine(Action<Options> options)
             ClrTypeConverter = new DefaultTypeConverter(this);
             BreakPoints = new List<BreakPoint>();
             DebugHandler = new DebugHandler(this);
+
+            BuildExtensionMethodTypeCache();
+            RegisterCachedExtensionMethods();
+        }
+
+        internal Dictionary<Type, List<MethodInfo>> ExtensionMethodTypeCache = new Dictionary<Type, List<MethodInfo>>();
```

Comment:
The solution is to have two MethodInfo dictionaries. One that is static and reused across engine instances (a cache) and one that contains the available methods from this engine instances.
So you build the second one by looking if the method infos are already available in the cache, then copy a reference to your engine local dictionary. If not you add it in both.
And to preserve thread-safety you `lock` the code that reads and writes to the cache dictionary.